pkg/endpoints/pipeline_test.go

```diff
@@ -14,1408 +14,1039 @@ limitations under the License.
+					corev1.Container{
+						Name: "Container2",
+					},
+				},
+			},
+		},
 	}
-	Container7 := corev1.Container{
-		Name: "Container7",
+	for _, pod := range pods {
+		_, err := r.K8sClient.CoreV1().Pods(namespace).Create(&pod)
+		if err != nil {
+			t.Errorf("Error creating pod '%s': %v\n", pod.Name, err)
+		}
 	}
 
-	TaskRun1 := v1alpha1.TaskRun{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: "TaskRun1",
-		},
-		Spec: v1alpha1.TaskRunSpec{},
-		Status: v1alpha1.TaskRunStatus{
-			PodName: "Pod1",
-		},
-	}
-	TaskRun2 := v1alpha1.TaskRun{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: "TaskRun2",
-		},
-		Spec: v1alpha1.TaskRunSpec{},
-		Status: v1alpha1.TaskRunStatus{
-			PodName: "Pod2",
+	taskRuns := []v1alpha1.TaskRun{
+		v1alpha1.TaskRun{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "TaskRun1",
+			},
+			Spec: v1alpha1.TaskRunSpec{},
+			Status: v1alpha1.TaskRunStatus{
+				PodName: pods[0].Name,
+			},
 		},
 	}
-	TaskRun3 := v1alpha1.TaskRun{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: "TaskRun3",
-		},
-		Spec: v1alpha1.TaskRunSpec{},
-		Status: v1alpha1.TaskRunStatus{
-			PodName: "Pod3",
-		},
+	for _, taskRun := range taskRuns {
+		_, err := r.PipelineClient.TektonV1alpha1().TaskRuns(namespace).Create(&taskRun)
+		if err != nil {
+			t.Errorf("Error creating taskRun '%s': %v\n", taskRun.Name, err)
+		}
 	}
 
-	Pod1 := corev1.Pod{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: "Pod1",
-		},
-		Spec: corev1.PodSpec{
-			Containers:     []corev1.Container{Container1, Container2},
-			InitContainers: []corev1.Container{Container3, Container4},
-		},
+	// Construct pipelineRunTaskRunStatus from taskRun Status
+	pipelineRunTaskRunStatus := map[string]*v1alpha1.PipelineRunTaskRunStatus{}
+	for i, _ := range taskRuns {
+		key := fmt.Sprintf("TaskRunStatus%d", i)
+		// The tasks within this pipelineRunTaskRunStatus do no exist
+		pipelineRunTaskRunStatus[key] = &v1alpha1.PipelineRunTaskRunStatus{
+			PipelineTaskName: fmt.Sprintf("Task%d", i),
+			Status:           &taskRuns[i].Status,
+		}
 	}
-	Pod2 := corev1.Pod{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: "Pod2",
-		},
-		Spec: corev1.PodSpec{
-			Containers:     []corev1.Container{Container5},
-			InitContainers: []corev1.Container{Container6},
+
+	pipelineName := "PipelineRun1"
+	pipelineRuns := []v1alpha1.PipelineRun{
+		v1alpha1.PipelineRun{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: pipelineName,
+			},
+			Spec: v1alpha1.PipelineRunSpec{},
+			Status: v1alpha1.PipelineRunStatus{
+				TaskRuns: pipelineRunTaskRunStatus,
+			},
 		},
 	}
-	Pod3 := corev1.Pod{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: "Pod3",
-		},
-		Spec: corev1.PodSpec{
-			Containers:     []corev1.Container{Container7},
-			InitContainers: []corev1.Container{},
-		},
+	for _, pipelineRun := range pipelineRuns {
+		_, err := r.PipelineClient.TektonV1alpha1().PipelineRuns(namespace).Create(&pipelineRun)
+		if err != nil {
+			t.Errorf("Error creating pipelineRun '%s': %v\n", pipelineRun.Name, err)
```

Comment:
I've taken to using t.Fatalf() because it stops the test run at that point, making it easier to use logs to work out what the first point of failure was. If a test has all ready failed, there's really no point in continuing. Would such a change fit into this overhaul?

Reply:
There are definitely certain places I could definitely be using fatal. For example, if creating the prerequisite CRD for the test fails (shouldn't happen, but still), I could return fatal. However, the create/update pipelineRun logic I grouped together because before it seems hard to grasp since it was broken up and there was a lot of overlap (in the set up). Even prior, getOne/getAll functionality was grouped together and I thought it made sense that testing an individual function should be scope to an individual test. Everything could be replaced to fatal, but it would not allow you to get the response for the rest of the functionality (e.g. `getX passed for all scenarios except Y` vs `getX failed early`). In either scenario, a fail is a fail and it shouldn't be approved, but if you prefer that style I could replaceAll?

Reply:
it's fine, I hadn't realised that you'd spent the time to merge them together.